Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+* [TBD] Number of resources (service monitors, pod monitors, probes and Prometheus rules) selected/rejected by the operator.
+
+```golang
+type PrometheusStatus struct {
+	// Represents whether any actions on the underlying managed objects are
+	// being performed. Only delete actions will be performed.
+	Paused bool `json:"paused"`
+	// Total number of non-terminated pods targeted by this Prometheus deployment
+	// (their labels match the selector).
+	Replicas int32 `json:"replicas"`
+	// Total number of non-terminated pods targeted by this Prometheus deployment
+	// that have the desired version spec.
+	UpdatedReplicas int32 `json:"updatedReplicas"`
+	// Total number of available pods (ready for at least minReadySeconds)
+	// targeted by this Prometheus deployment.
+	AvailableReplicas int32 `json:"availableReplicas"`
+	// Total number of unavailable pods targeted by this Prometheus deployment.
+	UnavailableReplicas int32 `json:"unavailableReplicas"`
+	// The current state of the Prometheus service.
+	// +patchMergeKey=type
+	// +patchMergeStrategy=merge
+	// +optional
+	Conditions []PrometheusStatusCondition `json:"conditions,omitempty"`
+	// The list has one entry per shard. Each entry provides a summary of the shard status.
+	// +patchMergeKey=shardID
+	// +patchMergeStrategy=merge
+	// +optional
+	ShardStatuses []ShardStatus `json:"shardStatuses,omitempty"`
+}
+
+// PrometheusStatusCondition represents the state of the resources associated with the Prometheus resource.
+// +k8s:deepcopy-gen=true
+type PrometheusStatusCondition struct {
+	// Type of the condition being reported.
+	// +required
+	Type PrometheusConditionType `json:"type"`
+	// status of the condition.
+	// +required
+	Status PrometheusConditionStatus `json:"status"`
+	// lastTransitionTime is the time of the last update to the current status property.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+	// Reason for the condition's last transition.
+	// +optional
+	Reason *string `json:"reason,omitempty"`
+	// Human-readable message indicating details for the condition's last transition.
+	// +optional
+	Message *string `json:"reason,omitempty"`
+}
+
+type PrometheusConditionType string
+
+const (
+	// Available indicates that enough Prometheus pods are ready to provide the service.
+	PrometheusAvailable PrometheusConditionType = "Available"
+	// Degraded indicates that some Prometheus pods don't run as expected.
+	PrometheusDegraded PrometheusConditionType = "Degraded"
+	// Reconciled indicates that the operator has reconciled the state of the underlying resources with the Prometheus object spec.
+	PrometheusReconciled PrometheusConditionType = "Reconciled"
+)
+
+type PrometheusConditionStatus string
+
+const (
+	PrometheusConditionTrue    PrometheusConditionStatus = "True"
+	PrometheusConditionFalse   PrometheusConditionStatus = "False"
+	PrometheusConditionUnknown PrometheusConditionStatus = "Unknown"
+)
```

Comment:
Can we use upstream [`(k8s.io/api/core/v1).ConditionStatus`](https://github.com/kubernetes/api/blob/master/core/v1/types.go#L2510-L2520) for this?

Reply:
We could. Now reading the v1.ConditionStatus comment, it says:

```
// These are valid condition statuses. "ConditionTrue" means a resource is in the condition.
// "ConditionFalse" means a resource is not in the condition. "ConditionUnknown" means kubernetes
// can't decide if a resource is in the condition or not. In the future, we could add other
// intermediate conditions, e.g. ConditionDegraded.
```

I wonder if we shouldn't merge the "Available" and "Degradeed" conditions types + introduce a "Degraded" condition status:
* Available/True => everything's running as expected
* Available/Degraded => some pods not in the correct state
* Available/False => at least 1 shard is down